[feature-wip] (memory tracker) (step2) Hook TCMalloc new/delete automatically counts to MemTracker

xinyiZzz opened the pull request:
# Proposed changes

Issue Number: close #7196 (step 2/3)

## Problem Summary:

Early Design Documentation: https://shimo.im/docs/DT6JXDRkdTvdyV3G

Implement a new way of memory statistics based on TCMalloc New/Delete Hook, MemTracker and TLS, and it is expected that all memory new/delete/malloc/free of the BE process can be counted.

## Checklist(Required)

1. Does it affect the original behavior: (Yes)
2. Has unit tests been added: (Yes)
3. Has document been added or modified: (Yes)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

### specific purpose:
#### 1. Accurate query memory limit.
- Before, exec_mem_limit is actually the memory limit at the Fragment Instance level, not the query. After, all Fragment Instance mem trackers of a query share a common ancestor query mem tracker, which will be a real memory limit for a query.
- After attaching to the query through tls, save the queryID, instanceID, and query mem tracker to tls. If the limit is exceeded when consume query mem tracker in the new/delete hook, the query will be canceled.
- All threads involved in a query runtime should attach the query through tls at startup, and detach the query when the thread exits, by using preset macros.
- Note that this may cause some previously successful queries to fail oom, and the query needs to be rewritten by hint set exec_mem_limit.

#### 2. Clearer tree structure of mem tracker.
- The mem tracker is clearly divided into process - query(task) pool - query(task) - fragment instance - exec node - exprs/hash table and others from top to bottom.
- When creating a tracker, the current tracker in tls is used as the parent by default, so the hierarchical relationship of mem tracker is equivalent to the hierarchical relationship of code, which avoids complicated mem tracker parameter passing.
- Because before, if you want to record the memory consumption of a location in the specified mem tracker, you need to pass the mem tracker layer by layer to the function of this location, which looks messy.
- After that, you only need to attach query or switch mem tracker externally, and you can get this mem tracker from tls at any internal location.
- This is especially useful for RowBatch and MemPool, the memory in RowBlock2 and RowBatch will be recorded directly in the tls tracker, avoiding a large number of mem trackers being passed as parameters, and creating a new temporary tracker in each Block/Batch, a huge number and useless.

#### 3. Complete and detailed BE memory statistics.
- Automatic memory statistics based on hook tcmalloc new/delete replaces the previous manual call consume/release mem tracker, which theoretically will not be missed.
- The logic of manually calling consume/release of specific memory is still reserved, but these mem trackers are created as virtual trackers, and cosume/release will not sync to parent, which is only used to improve the observability of running details.
Note, It is independent of the recording of tcmalloc hook in the thread local tracker, so the same block of memory is recorded independently in these two trackers.
Note, almost all the previous manual statistics positions are currently reserved, and some places may not be necessary, and will be revised in the future (**TODO**)
- if an independent memory allocator is used in a third-party library, special treatment is required. (**TODO**)
- How the cache is handled. In the LruCache insert phase, all rights of memory are transferred from the tracker held to the LruCache tracker. In the LruCache find phase, the transfer is reversed, and other caches are processed in the same way. We need a way to find where all the caches are. Implementing a memory-colored detection mode with reference to ASAN may be required (**TODO**)

### The difference between virutal tracker and non-virutal tracker:

- non-virutal tracker
In order to ensure the absolute accuracy of non-virutal mem tracker tree statistics, there are only two ways to count: one is to modify the tls mem tracker through attach or switch, and count in the tcmalloc new/delete hook; the other is to transfer memory ownership between non-virutal trackers.

- virutal tracker
Manual consume/release as before, the reasons for designing the virutal tracker: First, to transfer memory ownership between two trackers, it will release first and then consume, which is slower than calling consume/release directly on the virutal tracker; second, through parameters After blocking the virutal tracker, it will prevent the mem tracker tree from becoming more messy, and it is safer to add or delete the virutal tracker.

- The non-virutal tracker is similar to the INFO log level, and the virutal tracker is similar to the DEBUG log level.

### Existing performance problems and solutions:

- The mem tracker shared between threads has low performance when it is frequently consumed/released in the new/delete hook;
The memory consumption of the current thread is cached in tls, and after the cumulative consumption reaches 2M, the consume / release mem tracker will be called once to avoid frequent calls.

- At this stage, std::shared_ptr is used to save the mem tracker in tls. When a thread frequently switches the mem tracker, the use count of std::shared_ptr is frequently changed and the performance is low;
During an attach query, tls caches all mem trackers that have been switched and uncommitted memory consumption, and does not need to reset ptr when switching to the same mem tracker next time. In the future, the mem tracker in tls should be changed to raw pointers to solve this problem from the source (**TODO**)

### In the next pr
more detailed memory statistics such as exec node, exprs, hash table, etc. will be realized through the mem tracker switch during the thread attach query.

### Accuracy verification
1. Accurate process statistics
If the following three values are the same, the overall statistics are accurate:
- be_ip:webserver_port/mem_tracker - Process tracker - Current Consumption
- be_ip:webserver_port/memz - Mem Consumption
- top -p be_process_id

2. Query memory can be successfully limited
Modify session variable `set exec_mem_limit = 2147`, submit a query, it will return full OOM details: `Memory exceed limit. fragment=xxx details=xxx., on backend=xxx. Memory left in process limit=xxx. current tracker <label =xxx, used=xxx, limit=xxx, failed alloc size=xxx>.`

3. View more detailed memory statistics
Modify session variable `set mem_tracker_level = 2`, you can see INSTANCE level statistics, ref mem_tracker_level.

### Pformance Testing
The performance is reduced by about 1%~2% after opening the Hook TCMalloc new/delete.
Test Set:  ssb   LINEORDER    600w
Result:  
1. jmeter thread=1
- master, Avg: 559 (ms)
- track_new_delete=false, Avg: 558 (ms)
- track_new_delete=true, untracked_mem_limit_mbytes=2M,  Avg:  566 (ms)
- track_new_delete=true, untracked_mem_limit_mbytes=512k,  Avg: 565 (ms)
2. jmeter thread=10
- master, Avg: 3853 (ms)
- track_new_delete=false, Avg: 3834 (ms)
- track_new_delete=true, untracked_mem_limit_mbytes=2M,  Avg:  3929 (ms)
- track_new_delete=true, untracked_mem_limit_mbytes=512k,  Avg:  3912 (ms) 

Test SQL:
```
select k1, count(1) from  ( select LO_ORDERKEY as k1, count(1), max(LO_CUSTKEY)  from LINEORDER2 group by LO_ORDERKEY union all  select LO_CUSTKEY as k1, count(1), max(LO_ORDERKEY) from LINEORDER2 group by LO_CUSTKEY union all  select LO_ORDERKEY as k1, count(1), max(LO_CUSTKEY)  from LINEORDER2 group by LO_ORDERKEY union all  select LO_ORDERKEY as k1, count(1), max(LO_CUSTKEY)  from LINEORDER2 group by LO_ORDERKEY union all  select LO_CUSTKEY as k1, count(1), max(LO_ORDERKEY) from LINEORDER2 group by LO_CUSTKEY union all  select LO_ORDERKEY as k1, count(1), max(LO_CUSTKEY)  from LINEORDER2 group by LO_ORDERKEY union all  select LO_CUSTKEY as k1, count(1), max(LO_ORDERKEY) from LINEORDER2 group by LO_CUSTKEY union all  select LO_ORDERKEY as k1, count(1), max(LO_CUSTKEY)  from LINEORDER2 group by LO_ORDERKEY union all  select LO_CUSTKEY as k1, count(1), max(LO_ORDERKEY) from LINEORDER2 group by LO_CUSTKEY union all  select LO_CUSTKEY as k1, count(1), max(LO_ORDERKEY) from LINEORDER2 group by LO_CUSTKEY union all  select LO_ORDERKEY as k1, count(1), max(LO_CUSTKEY)  from LINEORDER2 group by LO_ORDERKEY union all  select LO_CUSTKEY as k1, count(1), max(LO_ORDERKEY) from LINEORDER2 group by LO_CUSTKEY union all  select LO_ORDERKEY as k1, count(1), max(LO_CUSTKEY)  from LINEORDER2 group by LO_ORDERKEY union all  select LO_ORDERKEY as k1, count(1), max(LO_CUSTKEY)  from LINEORDER2 group by LO_ORDERKEY union all  select LO_ORDERKEY as k1, count(1), max(LO_CUSTKEY)  from LINEORDER2 group by LO_ORDERKEY union all  select LO_CUSTKEY as k1, count(1), max(LO_ORDERKEY) from LINEORDER2 group by LO_CUSTKEY ) a group by k1 limit 10;
```

xinyiZzz committed:
hook_tcmalloc_count_mem_tracker

dataroaring commented on be/src/olap/task/engine_alter_tablet_task.cpp:
```diff
@@ -25,7 +25,14 @@ namespace doris {
-        : _alter_tablet_req(request) {}
+        : _alter_tablet_req(request) {
+    _mem_tracker = MemTracker::create_tracker(
+            config::memory_limitation_per_thread_for_schema_change * 1024 * 1024 * 1024,
```
we should put unit in config name, e.g. memory_limitation_per_thread_for_schema_change_gb

xinyiZzz replied:
done, thats right

dataroaring reviewed (commented):
The patch is very uncomfortable for reviewer due to many files it changed. I looked at about 100 files and added a few comment. 

You'd better split changes into small parts. e.g. 1. implement new memory tracker by hooking new, delete but not enable; 2. enable new memory tracker by removing old usage; 3. handle null pointer for old fasion allocate  4. add memory tracker for untracked memory usage;

This way, we can review the 2nd patch quickly and review others carefully.  The majority can be put into to the 2nd part, right?

xinyiZzz committed:
fix cache tracker

xinyiZzz committed:
fix comment

morningman commented:
> You'd better split changes into small parts. e.g. 1. implement new memory tracker by hooking new, delete but not enable; 2. enable new memory tracker by removing old usage; 3. handle null pointer for old fasion allocate 4. add memory tracker for untracked memory usage;

This is indeed a worse way to submit a PR, but for this current PR, I would suggest leaving it as is for now to avoid some more uncontrollable factors when splitting the PR.
We will gradually ask to split it as much as possible for subsequent PRs.
But DSIP is needed.

morningman commented on be/src/runtime/row_batch.cpp:
```diff
@@ -39,8 +40,8 @@ namespace doris {
-RowBatch::RowBatch(const RowDescriptor& row_desc, int capacity, MemTracker* mem_tracker)
-        : _mem_tracker(mem_tracker),
+RowBatch::RowBatch(const RowDescriptor& row_desc, int capacity)
+        : _mem_tracker(thread_local_ctx.get()->_thread_mem_tracker_mgr->mem_tracker()),
```
Is it possible that this rowbatch is transferred from one thread to another.
If so, the _mem_tracker also need to be changed?

xinyiZzz commented:
> The patch is very uncomfortable for reviewer due to many files it changed. I looked at about 100 files and added a few comment.
> 
> You'd better split changes into small parts. e.g. 1. implement new memory tracker by hooking new, delete but not enable; 2. enable new memory tracker by removing old usage; 3. handle null pointer for old fasion allocate 4. add memory tracker for untracked memory usage;
> 
> This way, we can review the 2nd patch quickly and review others carefully. The majority can be put into to the 2nd part, right?

thx for your suggestion, I will pay attention to control the size of pr later,
I consider moving some minor changes to the pr behind.
In fact, I have split a big feature into three, this is the second one...

xinyiZzz commented on be/src/runtime/thread_context.h:
```diff
@@ -50,27 +56,65 @@ class ThreadContext {
+            return "COMPACTION";
+        default:
+            return "UNKNOWN";
+        }
```
I found switch case to be faster, is using map more readable than using map? or other benefits.
Reference: https://stackoverflow.com/questions/6860525/c-what-is-faster-lookup-in-hashmap-or-switch-statement

dataroaring replied:
I means array not map. e.g.  const char * xx_desc[TYPE_NUM].  But maybe compiler translates such switch case to code like array?

xinyiZzz replied:
I see, it seems that compilers does compile switch constructs into array of code pointers, as said in the link above.

I changed to array as you said, which seems more concise.

Also, I did a simple test in the benchmark and it seems that switch constructs are faster, I didn't analyze the assembly code carefully, performance is not the bottleneck in this case.

![image](https://user-images.githubusercontent.com/13197424/158992378-7b821c8f-e528-4cca-b447-7f33a8f425e2.png)
![image](https://user-images.githubusercontent.com/13197424/158992899-70740800-6299-471f-a1c3-162fb8ea69ea.png)

xinyiZzz commented on be/src/runtime/thread_context.h:
```diff
@@ -50,27 +56,65 @@ class ThreadContext {
+    }
+
+    void consume_mem(int64_t size) {
+        if (thread_mem_tracker_mgr_init == true) {
```
done

xinyiZzz commented on be/src/exec/base_scanner.cpp:
```diff
@@ -39,14 +39,7 @@ BaseScanner::BaseScanner(RuntimeState* state, RuntimeProfile* profile,
-                  state->instance_mem_tracker())),
-#endif
-          _mem_pool(_mem_tracker.get()),
+          _mem_pool("BaseScanner:" + std::to_string(state->load_job_id())),
```
done

xinyiZzz commented on be/src/exec/es_http_scanner.cpp:
```diff
@@ -43,14 +42,7 @@ EsHttpScanner::EsHttpScanner(RuntimeState* state, RuntimeProfile* profile, Tuple
-                                            state->instance_mem_tracker())),
-#endif
-          _mem_pool(_mem_tracker.get()),
+          _mem_pool("EsHttpScanner:" + std::to_string(state->load_job_id())),
```
done

xinyiZzz commented on be/src/exec/tablet_info.cpp:
```diff
@@ -416,7 +416,7 @@ VOlapTablePartitionParam::VOlapTablePartitionParam(std::shared_ptr<OlapTableSche
           _t_param(t_param),
           _slots(_schema->tuple_desc()->slots()),
-          _mem_tracker(MemTracker::create_tracker(-1, "OlapTablePartitionParam")) {
+          _mem_tracker(MemTracker::create_virtual_tracker(-1, "OlapTablePartitionParam")) {
```
Because this tracker needs to manually consume/release
The difference between virutal tracker and non-virutal tracker:

- non-virutal tracker
In order to ensure the absolute accuracy of non-virutal mem tracker tree statistics, there are only two ways to count: one is to modify the tls mem tracker through attach or switch, and count in the tcmalloc new/delete hook; the other is to transfer memory ownership between non-virutal trackers.

-virutal tracker
Manual consume/release as before, the reasons for designing the virutal tracker: First, to transfer memory ownership between two trackers, it will release first and then consume, which is slower than calling consume/release directly on the virutal tracker; second, through parameters After blocking the virutal tracker, it will prevent the mem tracker tree from becoming more messy, and it is safer to add or delete the virutal tracker.
The non-virutal tracker is similar to the INFO log level, and the virutal tracker is similar to the DEBUG log level.

xinyiZzz commented on be/src/exec/tablet_sink.cpp:
```diff
@@ -48,6 +49,7 @@ NodeChannel::NodeChannel(OlapTableSink* parent, IndexChannel* index_channel, int
     if (_parent->_transfer_data_by_brpc_attachment) {
         _tuple_data_buffer_ptr = &_tuple_data_buffer;
     }
+    _node_channel_tracker = MemTracker::create_tracker(-1, "NodeChannel");
```
The BE id of a BE mem tracker is the same = _ =, so I added the thread id.

xinyiZzz commented on be/src/exprs/agg_fn_evaluator.cpp:
```diff
@@ -154,7 +154,7 @@ Status AggFnEvaluator::prepare(RuntimeState* state, const RowDescriptor& desc, M
 
     _string_buffer_len = 0;
-    _mem_tracker = mem_tracker;
+    _mem_tracker = MemTracker::create_virtual_tracker(-1, "AggFnEvaluator", mem_tracker);
```
The `memtracker` param is used in Expr::prepare on the next line, which I modified.
In addition, it is also the parent of the virtual tracker `_mem_tracker`.

xinyiZzz commented on be/src/olap/memtable.cpp:
```diff
@@ -31,14 +31,13 @@ namespace doris {
           _slot_descs(slot_descs),
           _keys_type(keys_type),
-          _mem_tracker(MemTracker::create_tracker(-1, "MemTable", parent_tracker)),
+          _mem_tracker(MemTracker::create_tracker(-1, "MemTable")),
```
The difference between virtual and non-virtual trackers is explained above

xinyiZzz commented on be/src/olap/rowset/segment_v2/segment.cpp:
```diff
@@ -51,10 +51,10 @@ Segment::Segment(const FilePathDesc& path_desc, uint32_t segment_id,
         : _path_desc(path_desc), _segment_id(segment_id), _tablet_schema(tablet_schema) {
 #ifndef BE_TEST
-    _mem_tracker = MemTracker::create_tracker(
+    _mem_tracker = MemTracker::create_virtual_tracker(
```
The difference between virtual and non-virtual trackers is explained above

xinyiZzz commented on be/src/olap/snapshot_manager.h:
```diff
@@ -99,6 +102,9 @@ class SnapshotManager {
     uint64_t _snapshot_base_id;
+
+    // TODO(zxy) used after
+    std::shared_ptr<MemTracker> _mem_tracker = nullptr;
```
In the next pr, this tracker will be switched to the tls mem tracker in the public func of `SnapshotManager`.

In this pr, I created all the trackers that will be used in the future, and built a complete mem tracker tree. (Perhaps it would be better to do this with a pr alone... = =||| )

xinyiZzz commented on be/src/olap/tablet_manager.cpp:
```diff
@@ -73,7 +73,7 @@ static bool _cmp_tablet_by_create_time(const TabletSharedPtr& a, const TabletSha
 
 TabletManager::TabletManager(int32_t tablet_map_lock_shard_size)
-        : _mem_tracker(MemTracker::create_tracker(-1, "TabletManager", nullptr,
+        : _mem_tracker(MemTracker::create_virtual_tracker(-1, "TabletManager", nullptr,
```
Because of the cache involved, the memory statistics of `TabletManager` were skipped before,
This has been implemented in the updated code.

xinyiZzz commented on be/src/olap/task/engine_alter_tablet_task.cpp:
```diff
@@ -25,7 +25,14 @@ namespace doris {
 EngineAlterTabletTask::EngineAlterTabletTask(const TAlterTabletReqV2& request)
-        : _alter_tablet_req(request) {}
+        : _alter_tablet_req(request) {
+    _mem_tracker = MemTracker::create_tracker(
```
Same as above, // TODO(zxy) used after

xinyiZzz commented on be/src/runtime/dpp_sink.cpp:
```diff
@@ -0,0 +1,926 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
no, need to be deleted
This file was deleted on the master, I didn't pay attention when I rebase

morningman reviewed (dismissed):
LGTM

xinyiZzz committed:
last review

xinyiZzz committed:
fix ut debug

morningman reviewed (approved):
LGTM

BiteTheDDDDt commented:
This pr seems will make build ASAN fail.

morningman commented:
> This pr seems will make build ASAN fail.

@yangzhg PTAL

xinyiZzz commented:
> This pr seems will make build ASAN fail.

I reproduced the problem in asan and it has been fixed in the following pr
#8569
